source/convergence_struct.cc

```diff
@@ -202,6 +202,214 @@ void ConvergenceAnalysisData<dim>::write_text(std::string filename) const
   }
 }
 
+namespace ConvergenceTest
+{
+
+namespace internal
+{
+  constexpr char header[] = "+------------------------------------------+"
+                        "----------------------+";
+
+  constexpr size_t column_width[2] ={ 40, 20 };
+
+  constexpr size_t line_width = 63;
+
+  template<typename Stream, typename A>
+  void add_line(Stream  &stream,
+                const A &line)
+  {
+    stream << "| "
+           << std::setw(line_width)
+           << line
+           << " |"
+           << std::endl;
+  }
+
+  template<typename Stream, typename A, typename B>
+  void add_line(Stream  &stream,
+                const A &first_column,
+                const B &second_column)
+  {
+    stream << "| "
+           << std::setw(column_width[0]) << first_column
+           << " | "
+           << std::setw(column_width[1]) << second_column
+           << " |"
+           << std::endl;
+  }
+
+  template<typename Stream>
+  void add_header(Stream  &stream)
+  {
+    stream << std::left << header << std::endl;
+  }
+
+} // internal
+
+
+ConvergenceTestParameters::ConvergenceTestParameters()
+:
+test_type(ConvergenceTestType::temporal),
+n_spatial_cycles(2),
+step_size_reduction_factor(0.5),
+n_temporal_cycles(2)
+{}
+
+
+ConvergenceTestParameters::ConvergenceTestParameters
+(const std::string &parameter_filename)
+:
+ConvergenceTestParameters()
+{
+  ParameterHandler prm;
+
+  declare_parameters(prm);
+
+  std::ifstream parameter_file(parameter_filename.c_str());
+
+  if (!parameter_file)
+  {
+    parameter_file.close();
+
+    std::ostringstream message;
+    message << "Input parameter file <"
+            << parameter_filename << "> not found. Creating a"
+            << std::endl
+            << "template file of the same name."
+            << std::endl;
+
+    std::ofstream parameter_out(parameter_filename.c_str());
+    prm.print_parameters(parameter_out,
+                         ParameterHandler::OutputStyle::Text);
+
+    AssertThrow(false, ExcMessage(message.str().c_str()));
+  }
+
+  prm.parse_input(parameter_file);
+
+  parse_parameters(prm);
+}
+
+
+void ConvergenceTestParameters::declare_parameters(ParameterHandler &prm)
+{
+  prm.enter_subsection("Convergence test parameters");
+  {
+    prm.declare_entry("Convergence test type",
+                      "temporal",
+                      Patterns::Selection("spatial|temporal|both"));
```

Comment:
`both` is defined here but is not implemented in the code.

Reply:
> `both` is defined here but is not implemented in the code.

This option is catched here

https://github.com/j507/RotatingMHD/blob/6f3db9e1c469afc24b5f2c18165f7258e5c17c12/source/convergence_struct.cc#L345-L362

In this PR, the new parameter structure is added but the way the convergence tests are performed is still *old*.

Reply:
In order to capture the spatio-temporal case, I added an assert in 36145cb.